plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -624,6 +628,33 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
                 .build();
     }
 
+    @Override
+    public void validateScan(ConnectorSession session, ConnectorTableHandle handle)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) handle;
+        Optional<PartitionSpec> partitionSpec = table.getPartitionSpecJson()
+                .map(partitionSpecJson -> PartitionSpecParser.fromJson(SchemaParser.fromJson(table.getTableSchemaJson()), partitionSpecJson));
+        if (partitionSpec.isEmpty() || partitionSpec.get().isUnpartitioned()) {
+            return;
+        }
+        if (isQueryPartitionFilterRequired(session)
+                && table.getEnforcedPredicate().equals(TupleDomain.all())) {
+            Set<String> columnsWithPredicates = table.getConstraintColumns().stream()
+                    .map(columnHandle -> ((IcebergColumnHandle) columnHandle).getName())
+                    .collect(toImmutableSet());
+            Set<String> partitionColumns = partitionSpec.get().fields().stream()
+                    .filter(field -> !(field.transform().isVoid()))
+                    .map(PartitionField::name)
+                    .collect(toImmutableSet());
+            if (!columnsWithPredicates.containsAll(partitionColumns)) {
+                String partitionColumnNames = String.join(", ", partitionColumns);
+                throw new TrinoException(
+                        StandardErrorCode.QUERY_REJECTED,
+                        format("Filter required on %s.%s for at least one partition column: %s", table.getSchemaName(), table.getTableName(), partitionColumnNames));
```

Comment:
Let's adjust the error message to say that filter is required on _all_ partition columns, and let the user know which one is missing